rust/agents/relayer/src/message_processor.rs

```diff
@@ -223,120 +238,146 @@ impl MessageProcessor {
+            .try_processing_message(latest_signed_checkpoint, message_leaf_index)
+            .await?
+        {
+            MessageProcessingStatus::Processed => {
+                self.processed_gauge.set(message_leaf_index as i64);
+                message_leaf_index + 1
+            }
+            MessageProcessingStatus::NotYetCheckpointed => {
+                // Do nothing. We should allow the backlog to be evaluated
+                // and will eventually learn about a new signed checkpoint.
+                message_leaf_index
+            }
+            MessageProcessingStatus::NotDestinedForInbox
+            | MessageProcessingStatus::NotWhitelisted => message_leaf_index + 1,
+            MessageProcessingStatus::Error => {
+                warn!(
+                    destination = self.inbox_contracts.inbox.local_domain(),
+                    leaf_index = message_leaf_index,
+                    "Message could not be processed, queue for retry"
+                );
+                self.retry_queue.push(MessageToRetry {
+                    leaf_index: message_leaf_index,
+                    time_to_retry: Reverse(Instant::now()),
+                    retries: 0,
+                });
+                message_leaf_index + 1
+            }
+        };
+        Ok(new_leaf)
+    }
+
+    /// Part of main loop
+    async fn retry_processing_messages(
+        &mut self,
+        latest_signed_checkpoint: &mut MultisigSignedCheckpoint,
+    ) -> Result<LoopControl> {
+        // See if we have messages to retry
+        if let Some(MessageToRetry { time_to_retry, .. }) = self.retry_queue.peek() {
+            // Since we use Reverse, we want time_to_retry to be smaller
+            if time_to_retry < &Reverse(Instant::now()) {
+                return Ok(Continue);
+            }
+        }
+        let MessageToRetry {
+            leaf_index,
+            retries,
+            ..
+        } = if let Some(v) = self.retry_queue.pop() {
+            v
+        } else {
+            // Nothing to do, just sleep
+            sleep(Duration::from_secs(1)).await;
+            return Ok(Flow);
+        };
+
+        info!(
+            destination = self.inbox_contracts.inbox.local_domain(),
+            leaf_index = leaf_index,
+            retries = retries,
+            retry_queue_length = self.retry_queue.len(),
+            "Retry processing of message"
+        );
+        match self
+            .try_processing_message(latest_signed_checkpoint, leaf_index)
+            .await?
+        {
+            MessageProcessingStatus::NotDestinedForInbox
+            | MessageProcessingStatus::NotYetCheckpointed
+            | MessageProcessingStatus::NotWhitelisted => {
+                error!(
+                    leaf_index = leaf_index,
+                    "Somehow we tried to retry a message that cant be retried"
+                );
+                bail!("Somehow we tried to retry a message that cant be retried")
```

Comment:
I feel `error!` should be reserved for when you decide to panic OR when you are recovering from it in a way which may degrade performance from the expected/norm. Otherwise I would say that at the site of a `bail!` you should stick to `info!` or `warn!` depending, and we could create a `bail_and_warn!` macro which does make some sense to me.

Reply:
I feel `error!` should be reserved for when you decide to panic OR when you are recovering from it in a way which may degrade performance from the expected/norm. Otherwise I would say that at the site of a `bail!` you should stick to `info!` or `warn!` depending, and we could create a `warn_and_bail!` macro which does make some sense to me.

Reply:
That said I am just keeping it as is for now